File: src/benchmarks/mandelbrot.js
// Mandelbrot Benchmark
// Author: Peter Jensen
(function () {

  // Kernel configuration
  var kernelConfig = {
    kernelName:       "Mandelbrot",
    kernelInit:       initMandelbrot,
    kernelSimd:       simdMandelbrot,
    kernelNonSimd:    nonSimdMandelbrot,
    kernelIterations: 10000
  };

  // Hook up to the harness
  benchmarks.add (new Benchmark (kernelConfig));

  function float32x4ToString (f4) {
    return "[" + f4.x + "," + f4.y + "," + f4.z + "," + f4.w + "]";
  }

  function uint32x4ToString (i4) {
    return "[" + i4.x + "," + i4.y + "," + i4.z + "," + i4.w + "]";
  }

  function mandelx1 (c_re, c_im, max_iterations) {
    var z_re = c_re,
        z_im = c_im,
        i;
    for (i = 0; i < max_iterations; i++) {
      var z_re2 = z_re*z_re;
      var z_im2 = z_im*z_im;
      if (z_re2 + z_im2 > 4.0)
        break;

      var new_re = z_re2 - z_im2;
      var new_im = 2.0 * z_re * z_im;
      z_re = c_re + new_re;
      z_im = c_im + new_im;
    }
    return i;
  }

  function mandelx4(c_re4, c_im4, max_iterations) {
    var z_re4  = SIMD.toFloat32x4 (c_re4);  // create a clone, rename to .clone() ???
    var z_im4  = SIMD.toFloat32x4 (c_im4);  // create a clone
    var four4  = float32x4.splat (4.0);
    var two4   = float32x4.splat (2.0);
    var count4 = uint32x4.splat (0);
    var one4   = uint32x4.splat (1);

    for (var i = 0; i < max_iterations; ++i) {
      var z_re24 = SIMD.mul (z_re4, z_re4);
      var z_im24 = SIMD.mul (z_im4, z_im4);

      var mi4    = SIMD.lessThanOrEqual (SIMD.add (z_re24, z_im24), four4);
      // if all 4 values are greater than 4.0, there's no reason to continue
      if (mi4.signMask === 0x00) {
        break;
      }

      var new_re4 = SIMD.sub (z_re24, z_im24);
      var new_im4 = SIMD.mul (SIMD.mul (two4, z_re4), z_im4);
      z_re4       = SIMD.add (c_re4, new_re4);
      z_im4       = SIMD.add (c_im4, new_im4);
      count4      = SIMD.add (count4, SIMD.and (mi4, one4));
    }
    return count4;
  }

  function initMandelbrot () {
    var simd    = simdMandelbrot (1);
    var nonSimd = nonSimdMandelbrot (1);
    if (simd.length !== nonSimd.length) {
      return false;
    }
    for (var i = 0, n = simd.length; i < n; ++i) {
      if (simd[i] !== nonSimd[i]) {
        return false;
      }
    }
    return true;
  }

  // Non SIMD version of the kernel
  function nonSimdMandelbrot (n) {
    var result = new Array (4);
    for (var i = 0; i < n; ++i) {
      result [0] = mandelx1 (0.0, 0.0, 100);
      result [1] = mandelx1 (0.0, 0.0, 100);
      result [2] = mandelx1 (0.0, 0.0, 100);
      result [3] = mandelx1 (0.0, 0.0, 100);
    }
    return result;
  }

  // SIMD version of the kernel
  function simdMandelbrot (n) {
    var result = new Array (4);
    var zero4  = float32x4.zero ();
    for (var i = 0; i < n; ++i) {
      var r = mandelx4 (zero4, zero4, 100);
      result [0] = r.x;
      result [1] = r.y;
      result [2] = r.z;
      result [3] = r.w;
    }
    return result;
  }

} ());

File: dart/bin/mandelbrot.dart
/*
  Copyright (C) 2013 John McCutchan <john@johnmccutchan.com>

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

*/

library vector_math_matrix_bench;

import 'dart:typed_data';
import 'package:vector_math/vector_math_operations.dart';
import 'package:benchmark_harness/benchmark_harness.dart';
import 'benchmark_contrast.dart';

class NonSIMDMandelbrotBenchmark extends BenchmarkBase {
  const NonSIMDMandelbrotBenchmark() : super("NonSIMDMandelbrot");

  static void main() {
    new NonSIMDMandelbrotBenchmark().report();
  }

  int mandelx1 (double c_re, double c_im, int max_iterations) {
    double z_re = c_re;
    double z_im = c_im;
    int    i;
    for (i = 0; i < max_iterations; i++) {
      double z_re2 = z_re*z_re;
      double z_im2 = z_im*z_im;
      if (z_re2 + z_im2 > 4.0)
        break;

      double new_re = z_re2 - z_im2;
      double new_im = 2.0 * z_re * z_im;
      z_re = c_re + new_re;
      z_im = c_im + new_im;
    }
    return i;
  }

  void run() {
    float c_real = -0.50075;
    float c_imagine = 0.003;
    int max_iterations = 100;
    for (int i = 0; i < 80000; i++) {
      mandelx1(c_real, c_imagine, max_iterations);
    }
  }
}

class SIMDMandelbrotBenchmark extends BenchmarkBase {
  const SIMDMandelbrotBenchmark() : super("SIMDMandelbrot");
  final Int32x4 one4 = new Int32x4(1, 1, 1, 1);

  static void main() {
    new SIMDMandelbrotBenchmark().report();
  }

  Int32x4 mandelx4(Float32x4 c_re4, Float32x4 c_im4, int max_iterations) {
    Float32x4 z_re4  = c_re4;
    Float32x4 z_im4  = c_im4;
    Float32x4 four4  = new Float32x4.splat(4.0);
    Float32x4 two4   = new Float32x4.splat(2.0);
    Int32x4 count4  = new Int32x4(0, 0, 0, 0);

    int i;
    bool done = false;
    for (i = 0; !done && i < max_iterations; ++i) {
      Float32x4 z_re24 = z_re4 * z_re4;
      Float32x4 z_im24 = z_im4 * z_im4;
      Int32x4 mi4 = (z_re24 + z_im24).lessThan(four4);
      done = mi4.signMask == 0x0;
      Float32x4 new_re4 = z_re24 - z_im24;
      Float32x4 new_im4 = two4 * z_re4 * z_im4;
      z_re4 = c_re4 + new_re4;
      z_im4 = c_im4 + new_im4;
      Int32x4 add01 = mi4 & one4;
      count4 = count4 + add01;
    }
    return count4;
  }

  void run() {
    Float32x4 c_real = new Float32x4(-0.50075, -0.50075, -0.50075, -0.50075);
    Float32x4 c_imagine = new Float32x4(0.003, 0.003, 0.003, 0.003);
    int max_iterations = 100;
    for (int i = 0; i < 20000; i++) {
      mandelx4(c_real, c_imagine, max_iterations);
    }
  }
}

class MandelbrotBenchmark {
  static Object create() {
    return new BenchmarkContrast("Mandelbrot",
                                 new NonSIMDMandelbrotBenchmark(),
                                 new SIMDMandelbrotBenchmark());
  }
}
